exchange/targeting.go

```diff
@@ -35,49 +36,80 @@ type targetData struct {
 // The one exception is the `hb_cache_id` key. Since our APIs explicitly document cache keys to be on a "best effort" basis,
 // it's ok if those stay in the auction. For now, this method implements a very naive cache strategy.
 // In the future, we should implement a more clever retry & backoff strategy to balance the success rate & performance.
-func (targData *targetData) setTargeting(auc *auction, isApp bool, categoryMapping map[string]string, truncateTargetAttr *int) {
+func (targData *targetData) setTargeting(auc *auction, isApp bool, categoryMapping map[string]string, truncateTargetAttr *int, multiBidMap ExtMultiBidMap, accountMaxBids int) {
 	for impId, topBidsPerImp := range auc.winningBidsByBidder {
 		overallWinner := auc.winningBids[impId]
-		for bidderName, topBidPerBidder := range topBidsPerImp {
-			isOverallWinner := overallWinner == topBidPerBidder
+		for originalBidderName, topBidsPerBidder := range topBidsPerImp {
+			targetingBidderCode := originalBidderName
 
-			targets := make(map[string]string, 10)
-			if cpm, ok := auc.roundedPrices[topBidPerBidder]; ok {
-				targData.addKeys(targets, openrtb_ext.HbpbConstantKey, cpm, bidderName, isOverallWinner, truncateTargetAttr)
-			}
-			targData.addKeys(targets, openrtb_ext.HbBidderConstantKey, string(bidderName), bidderName, isOverallWinner, truncateTargetAttr)
-			if hbSize := makeHbSize(topBidPerBidder.Bid); hbSize != "" {
-				targData.addKeys(targets, openrtb_ext.HbSizeConstantKey, hbSize, bidderName, isOverallWinner, truncateTargetAttr)
-			}
-			if cacheID, ok := auc.cacheIds[topBidPerBidder.Bid]; ok {
-				targData.addKeys(targets, openrtb_ext.HbCacheKey, cacheID, bidderName, isOverallWinner, truncateTargetAttr)
-			}
-			if vastID, ok := auc.vastCacheIds[topBidPerBidder.Bid]; ok {
-				targData.addKeys(targets, openrtb_ext.HbVastCacheKey, vastID, bidderName, isOverallWinner, truncateTargetAttr)
-			}
-			if targData.includeFormat {
-				targData.addKeys(targets, openrtb_ext.HbFormatKey, string(topBidPerBidder.BidType), bidderName, isOverallWinner, truncateTargetAttr)
+			var bidderCodePrefix string
+			maxBids := 1
+			if multiBid, ok := multiBidMap[originalBidderName.String()]; ok {
+				bidderCodePrefix = multiBid.TargetBidderCodePrefix
+				maxBids = *multiBid.MaxBids
 			}
 
-			if targData.cacheHost != "" {
-				targData.addKeys(targets, openrtb_ext.HbConstantCacheHostKey, targData.cacheHost, bidderName, isOverallWinner, truncateTargetAttr)
-			}
-			if targData.cachePath != "" {
-				targData.addKeys(targets, openrtb_ext.HbConstantCachePathKey, targData.cachePath, bidderName, isOverallWinner, truncateTargetAttr)
-			}
+			for i, topBid := range topBidsPerBidder {
+				if accountMaxBids != 0 && i >= accountMaxBids {
+					topBid.Bid = nil
```

Comment:
Looks like we're enforcing the max bids limit in this function. Doing this in `setTargeting` is kinda misleading IMHO. We should have a separate function where the bid limit is enforced. It might be slightly inefficient but it will make the code easier to read and understand. 

Also, at this point in the code we have already done some bid processing like setting up even tracking info, doing currency conversions, etc. Is there a way we can move this logic upstream so that we get rid of the bids earlier in the process and don't avoid processing extra bids that might be potentially discarded?

Reply:
We can certainly isolate this logic, but it should only be done after the auction has ended. We would want to eliminate only the "least wanted" or "losing bids.

related to https://github.com/prebid/prebid-server/pull/2468#discussion_r1099836817